Fix the static_functions_not_mutating error messsage

felixinkinto opened the pull request:
Fix https://github.com/swiftlang/swift/issues/77835

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1718,7 +1718,7 @@ ERROR(functions_mutating_and_not,none,
       (SelfAccessKind, SelfAccessKind))
 ERROR(static_functions_not_mutating,none,
-      "static functions must not be declared mutating", ())
+      "nonmutating or mutating modifiers cannot be used on static functions because they do not operate on instance properties or self", ())
```
Instead of clamping mutating and nonmutating together it would be better to print one or the other depending on what the attribute was, a boolean flag could be used to distinguish between the two (i.e. `no_overloads_match_exactly_in_call` does a similar thing reference vs. call). 

Instead of "operate" we can use "cannot access" and the "because" part can be turned into a note.

felixinkinto replied:
@xedin 
Thanks for the advice!
I fixed based on your suggestion.

xedin reviewed (changes requested):
Thank you for the contribution! I left a comment inline about the text and as a separate note: please modify existing tests (because the text is now different) or make sure that we have tests for both `nonmutating` and `mutating` attributes that exercise the new diagnostics before we can run CI and land this.

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1718,7 +1718,10 @@ ERROR(functions_mutating_and_not,none,
+      "%0 modifier cannot be used on static functions",
+      (SelfAccessKind))
+NOTE(static_functions_not_mutating_detail,none,
+      "%0 modifier cannot access instance properties or self", (SelfAccessKind))
```
I think this should be about `static members` not the modifier, right?

felixinkinto replied:
reworded as `static members cannot access instance properties or self`
https://github.com/swiftlang/swift/pull/78052/commits/e8b9b8cc69fe3e4a6ad0a059e6041176838045a0

xedin reviewed (changes requested):
I left one more small note inline. We also have a few tests that use the diagnostic and they'd need to be adjusted before we can land this:

test/Sema/immutability.swift:134
test/decl/subscript/static.swift:23
test/decl/subscript/static.swift:24
test/decl/overload.swift:413

xedin commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -603,7 +603,8 @@ void AttributeChecker::visitMutationAttr(DeclAttribute *attr) {
   if (FD->isStatic())
-    diagnoseAndRemoveAttr(attr, diag::static_functions_not_mutating);
+    diagnoseAndRemoveAttr(attr, diag::static_functions_not_mutating, attrModifier);
+    diagnose(attr->getLocation(), diag::static_functions_not_mutating_detail, attrModifier);
```
I think this point to `FD` itself.

felixinkinto replied:
i fixed to `diagnose(FD, diag::static_functions_not_mutating_detail);`
https://github.com/swiftlang/swift/pull/78052/commits/e8b9b8cc69fe3e4a6ad0a059e6041176838045a0

felixinkinto committed:
fix static_functions_not_mutating error msg

felixinkinto commented:
@xedin sorry for the delay for long. I addressed the feedback. May I ask some questions to do behaviour checks?
1. Does the order matter? I realized note usually follows the error in other areas.
```
➜  test xcrun -sdk macosx build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swiftc test.swift
test.swift:4:3: error: 'nonmutating' modifier cannot be used on static functions
2 |
3 | struct S {
4 |   nonmutating static func foo() {}
  |   |                       `- note: static members cannot access instance properties or self
  |   `- error: 'nonmutating' modifier cannot be used on static functions
5 |
6 |
```
2. After testing the locally built compiler in Xcode according to [this manual](https://github.com/swiftlang/swift/blob/main/docs/HowToGuides/FAQ.md#how-do-i-use-a-locally-built-compiler-to-build-x), it seems that the existing swiftc is overriding the errors in the IDE. 
<img width="304" alt="Screenshot 2025-03-07 at 10 22 06" src="https://github.com/user-attachments/assets/de317985-4157-41f4-adbe-93d430371318" /> this build log shows my local compiler.

<img width="306" alt="Screenshot 2025-03-07 at 10 22 17" src="https://github.com/user-attachments/assets/10eebf66-8673-4c8a-a67d-ae165fba47c0" /> this is still the default swiftc

felixinkinto commented:
> @felixinkinto Next time, please [ask us](https://github.com/swiftlang/swift/blob/55189bae8e5516967998000deaa0e138a1c9f4fa/docs/HowToGuides/FirstPullRequest.md#how-do-i-pick-something-to-work-on) to assign you to an issue before you start working on one. A good first issue is a queue, not a race. We want to avoid instances of multiple people independently working on them.

sorry I’ll make sure to ask for an assignment before starting work on an issue next time.

felixinkinto commented:
@swift-ci Please smoke test

felixinkinto commented:
@swift-ci please test Windows platform

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1735,7 +1735,10 @@ ERROR(functions_mutating_and_not,none,
       (SelfAccessKind, SelfAccessKind))
 ERROR(static_functions_not_mutating,none,
-      "static functions must not be declared mutating", ())
+      "%0 modifier cannot be used on static functions",
+      (SelfAccessKind))
```
If you incorporate the declaration like so, the message will be also be tailored for an accessor:
```suggestion
      "%0 modifier cannot be used on %kindonly1",
      (SelfAccessKind, const FuncDecl *))
```

We loose the plural, but I think that’s fine because the object remains indefinite.

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1735,7 +1735,10 @@ ERROR(functions_mutating_and_not,none,
-      "static functions must not be declared mutating", ())
+      "%0 modifier cannot be used on static functions",
+      (SelfAccessKind))
+NOTE(static_functions_not_mutating_detail,none,
+      "static members cannot access instance properties or self", ())
```
I’m not a fan of this interpretation because it’s misleading. We have this very problem with another diagnostic: https://github.com/swiftlang/swift/issues/62909. You can access both `self` or instance members inside a static method depending on what you mean by it.

Maybe we could borrow a more clear phrasing from the [book](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/methods#Type-Methods). For example:

```
static members are defined on the type itself rather than on instances of the type
static members belong to the type itself rather than to instances of the type
static members are members of the type itself rather than instances of the type
```

AnthonyLatsis commented:
> Does the order matter? I realized note usually follows the error in other areas.

Emission order does matter. Notes are always attached to a primary diagnostic such as an error or warning. The grouped diagnostic formatter probably chose to display the error below the note there because it wants a clear path from each message to its location.
